presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,440 @@
+    }
+
+    @VisibleForTesting
+    static class JoinEnumerator
+    {
+        private final Map<Set<PlanNode>, JoinEnumerationResult> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final Ordering<JoinEnumerationResult> resultOrdering;
+        private final EqualityInference allInference;
+        private final Expression allFilter;
+        private final SymbolAllocator symbolAllocator;
+        private final Lookup lookup;
+
+        @VisibleForTesting
+        JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup, Expression filter, CostComparator costComparator)
+        {
+            requireNonNull(idAllocator, "idAllocator is null");
+            requireNonNull(symbolAllocator, "symbolAllocator is null");
+            requireNonNull(session, "session is null");
+            requireNonNull(lookup, "lookup is null");
+            requireNonNull(filter, "filter is null");
+            requireNonNull(costComparator, "costComparator is null");
+            this.idAllocator = idAllocator;
+            this.symbolAllocator = symbolAllocator;
+            this.session = session;
+            this.lookup = lookup;
+            this.resultOrdering = getResultOrdering(costComparator, session);
+            this.allInference = createEqualityInference(filter);
+            this.allFilter = filter;
+        }
+
+        private static Ordering<JoinEnumerationResult> getResultOrdering(CostComparator costComparator, Session session)
+        {
+            return new Ordering<JoinEnumerationResult>()
+            {
+                @Override
+                public int compare(JoinEnumerationResult result1, JoinEnumerationResult result2)
+                {
+                    checkArgument(result1.cost.isPresent() && result2.cost.isPresent(), "costs not present");
+                    return costComparator.compare(session, result1.cost.get(), result2.cost.get());
+                }
+            };
+        }
+
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
+            JoinEnumerationResult bestResult = memo.get(multiJoinKey);
+            if (bestResult == null) {
+                checkState(sources.size() > 1);
+                List<JoinEnumerationResult> results = generatePartitions(sources.size())
+                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning))
+                        .collect(toImmutableList());
+                if (results.stream().anyMatch(result -> result.cost.isPresent() && result.cost.get().isUnknown())) {
+                    return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
```

Comment:
skipping.  We don't keep evaluating join orders once we get an unknown cost, so there's no reason to add it to the memo